### PR TITLE
Fixes composer installation errors due to problematic configuration files

### LINKED ISSUE #4: Package is triggering autoload errors
For some reason our modules are triggering class not found exceptions when this package is added. It looks like this is only triggered when installing packages from source (git)

```


  [Symfony\Component\Debug\Exception\FatalThrowableError]
  Class '<Redacted>\UserModule\Module' not found


Exception trace:
 () at /data/vendor/<redacted>/user-module/config/user.config.php:53
 include() at /data/vendor/<redacted>/user-module/config/module.config.php:23
 include() at /data/vendor/zfcampus/zf-asset-manager/src/AssetInstaller.php:74
 ZF\AssetManager\AssetInstaller->__invoke() at /data/vendor/zfcampus/zf-asset-manager/src/Plugin.php:65
 ZF\AssetManager\Plugin->onPostPackageInstall() at n/a:n/a
 call_user_func() at phar:///usr/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php:171
 Composer\EventDispatcher\EventDispatcher->doDispatch() at phar:///usr/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php:116
 Composer\EventDispatcher\EventDispatcher->dispatchPackageEvent() at phar:///usr/bin/composer/src/Composer/Installer.php:605
 Composer\Installer->doInstall() at phar:///usr/bin/composer/src/Composer/Installer.php:223
 Composer\Installer->run() at phar:///usr/bin/composer/src/Composer/Command/InstallCommand.php:119
 Composer\Command\InstallCommand->execute() at phar:///usr/bin/composer/vendor/symfony/console/Command/Command.php:267
 Symfony\Component\Console\Command\Command->run() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:846
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:191
 Symfony\Component\Console\Application->doRun() at phar:///usr/bin/composer/src/Composer/Console/Application.php:227
 Composer\Console\Application->doRun() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:122
 Symfony\Component\Console\Application->run() at phar:///usr/bin/composer/src/Composer/Console/Application.php:100
 Composer\Console\Application->run() at phar:///usr/bin/composer/bin/composer:54
 require() at /usr/bin/composer:24

````

### PR DESCRIPTION
**Pull request by weierophinney**
If a module configuration file contains elements that depend on autoloading and/or elements that are clearly not configuration, calling `include` on it can lead to failure to complete a composer installation/removal action.

This patch modifies the logic as follows:

- If a configuration file does not have `asset_manager` configuration, it is skipped.
- Both installation and uninstallation actions tokenize the configuration file to look for problematic elements, notably `exec()` and `eval()` calls; these raise warnings by the installer if detected.
- Instead of performing install operations immediately, the `Plugin` class now aggregates install operations as closures around the relevant events, and executes these during a `post-autoload-dump` plugin hook; this fixes autoloading issues. Uninstall operations still occur immediately as the relevant uninstall and update events.

Fixes #4

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -47,18 +48,22 @@ matrix:
     - php: hhvm
 
 before_install:
-  - if [[ $TRAVIS_PHP_VERSION != "hhvm" && $TEST_COVERAGE != 'true' ]]; then phpenv config-rm xdebug.ini ; fi
+  - if [[ $TEST_COVERAGE != 'true' ]]; then phpenv config-rm xdebug.ini || return 0 ; fi
+  - travis_retry composer self-update
```
Comment (michalbundyra) [inline .travis.yml]: This line is redundant. Travis update composer automatically, please see:
https://travis-ci.org/zfcampus/zf-asset-manager/jobs/296348988#L473
https://travis-ci.org/zfcampus/zf-asset-manager/jobs/296348988#L478

File: src/UnparseableTokensTrait.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        T_EVAL,
+        T_EXIT,
+        T_EXTENDS,
+        T_INTERFACE,
+        T_NEW,
+        T_TRAIT,
+    ];
+
+    /**
+     * @var string $packageConfigPath
```
Comment (michalbundyra) [inline src/UnparseableTokensTrait.php]: `@param` instead of `@var`

File: src/UnparseableTokensTrait.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    {
+        $contents = file_get_contents($packageConfigPath);
+        if (preg_match('/[\'"]asset_manager[\'"]\s*\=\>/s', $contents)) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @var string $packageConfigPath
```
Comment (michalbundyra) [inline src/UnparseableTokensTrait.php]: As above, `@param` instead of `@var`

File: src/Plugin.php [hunk 4]
```diff
@@ -55,21 +63,35 @@ public function activate(Composer $composer, IOInterface $io)
...
+            $installer($event);
+        };
     }
 
     /**
      * Updates assets provided by the package, if any.
      *
      * Uninstalls any previously installed assets for the package, and then
-     * runs an install for the package.
+     * memoizes an install operation to run post-autoload-dump.
```
Comment (Xerkus) [inline src/Plugin.php]: typo in memorizes

Comment (weierophinney) [inline src/Plugin.php]: No typo: read about [memoization](https://en.wikipedia.org/wiki/Memoization) on wikipedia. :smile:

File: src/UnparseableTokensTrait.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        $contents = file_get_contents($packageConfigPath);
+        $tokens = token_get_all($contents);
+        foreach ($tokens as $index => $token) {
+            if (! is_array($token)) {
+                continue;
+            }
+
+            // ::class notation is okay
+            if ($token[0] === T_DOUBLE_COLON
```
Comment (Xerkus) [inline src/UnparseableTokensTrait.php]: this check is no longer needed

File: src/UnparseableTokensTrait.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            }
+        }
+        return true;
+    }
+
+    /**
+     * @param string|array Token to test as a class pseudoconstant
+     * @return bool
+     */
+    private function isClassPseudoConstant($token)
```
Comment (Xerkus) [inline src/UnparseableTokensTrait.php]: same here, no longer needed

**Pull request merged**